# Euler Project Challenges

In this document I will be contributing solutions to challenges in the Euler project. I will mostly work though in order and will include links to the questions or copy out the question text.



In [2]:
##hello 
import numpy as np
import timeit 



## Problem 1
If we list all the natural numbers below 10 that are multiples of 3 or 5, we get 3, 5, 6 and 9. The sum of these multiples is 23.

Find the sum of all the multiples of 3 or 5 below 1000.

In [5]:
limit =100
a = 3
b = 5
factors = 0
factors_array = []

for i in range(limit):
    div_a = i%a
    div_b = i%b
    if div_a == 0:
        factors += i
        factors_array.append(i)
        continue
    elif div_b == 0:
        factors += i
        factors_array.append(i)
        continue

print(factors)
print(factors_array)

2318
[0, 3, 5, 6, 9, 10, 12, 15, 18, 20, 21, 24, 25, 27, 30, 33, 35, 36, 39, 40, 42, 45, 48, 50, 51, 54, 55, 57, 60, 63, 65, 66, 69, 70, 72, 75, 78, 80, 81, 84, 85, 87, 90, 93, 95, 96, 99]


Answer

2318

Answer = 233168


### Problem 2
Each new term in the Fibonacci sequence is generated by adding the previous two terms. By starting with 1 and 2, the first 10 terms will be:

1, 2, 3, 5, 8, 13, 21, 34, 55, 89, ...

By considering the terms in the Fibonacci sequence whose values do not exceed four million, find the sum of the even-valued terms.

In [10]:
limit = 4000000
fib_seq = [1,2]
fib_sum = 0

## Generating the fibonacci sequence up to the defined limit
for i in range(limit):
    fib_var = fib_seq[i] + fib_seq[i+1]
    if fib_var <= limit:
        fib_seq.append(fib_var)
    else: break
print(fib_seq)

## summing the even values in this range
for i in fib_seq:
    even_test = i%2
    if even_test == 0:
        fib_sum += i
print(fib_sum)

[1, 2, 3, 5, 8, 13, 21, 34, 55, 89, 144, 233, 377, 610, 987, 1597, 2584, 4181, 6765, 10946, 17711, 28657, 46368, 75025, 121393, 196418, 317811, 514229, 832040, 1346269, 2178309, 3524578]
4613732


Answer : 4613732

### Problem 3
The prime factors of 13195 are 5, 7, 13 and 29.

What is the largest prime factor of the number 600851475143 ?

#### Thoughts:
Working down from the number would provide factors but be a much slower way to iterate and may lead to ~1e12 operations given the size of the number.
Instead we will work upwards, finding all factors and then convert these into primie factors.
I suspect there is a nice solution for this online, after finding my own solution I will compare

In [6]:
# Time required to run x iterations where x is the number in which we are finding prime factors . This ran for >12mins so clearly too slow
print(timeit.timeit("counter += 1", "counter = 0", number=600851475143))

In [16]:
prime_parent = 600851475143
# as we find primes we will divide the original number by these primes to reduce the number of iterations needed
#  prime_parent_new servesas this variable
prime_parent_new = prime_parent
prime_factor_list = []
prime_factor_mult = 1
# i is our iterating vairable, we ignore 1 as this is not a prime number
i = 2

while prime_parent_new > 1:
    if prime_parent_new%i == 0:
        #if i is a factor and is unique in the factor list then append
        if i not in prime_factor_list: 
            prime_factor_list.append(i)
        prime_parent_new = prime_parent_new/i
    else:
        i += 1
print(prime_factor_list)
print(max(prime_factor_list))


        

[71, 839, 1471, 6857]
6857


Answer 6857 

Note: I was unable to find a cleaner solution online so will keep the above

### Problem 4

A palindromic number reads the same both ways. The largest palindrome made from the product of two 2-digit numbers is 9009 = 91 × 99.

Find the largest palindrome made from the product of two 3-digit numbers.


In [63]:
lowerbound = 100
upperbound = 999

palindromes = []

for i in range(upperbound,lowerbound,-1):
    for j in range(upperbound,lowerbound,-1):
        # convert to a string to be able to reverse the digits
        prod = str(i*j)
        prod_reverse = reversed(prod)
        if list(prod) == list(prod_reverse):
            pal_element = [int(prod),i,j]
            palindromes.append(pal_element)

b = np.array(palindromes)
max_val = max(b[:,0])

i = np.where(b[:,0] == max_val)
print(b[i,:])
# print(j)

[[[906609    993    913]
  [906609    913    993]]]


Answer 906609

The multiple of 913 and 993.
I'm not partivularly happy with the output here, I'd like to find a cleaner way to only extract the 993 & 913 and not the reverse which also gives the palindrom

Problem 5